core/drivers/stm32_rng.c

```diff
@@ -0,0 +1,188 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (c) 2018, STMicroelectronics - All Rights Reserved
+ */
+
+#include <assert.h>
+#include <drivers/stm32_rng.h>
+#include <io.h>
+#include <kernel/dt.h>
+#include <kernel/delay.h>
+#include <kernel/generic_boot.h>
+#include <kernel/panic.h>
+#include <mm/core_memprot.h>
+#include <stdbool.h>
+#include <stm32_util.h>
+#include <string.h>
+
+#ifdef CFG_DT
+#include <libfdt.h>
+#endif
+
+#define DT_RNG_COMPAT		"st,stm32-rng"
+#define RNG_CR			0x00U
+#define RNG_SR			0x04U
+#define RNG_DR			0x08U
+
+#define RNG_CR_RNGEN		BIT(2)
+#define RNG_CR_IE		BIT(3)
+#define RNG_CR_CED		BIT(5)
+
+#define RNG_SR_DRDY		BIT(0)
+#define RNG_SR_CECS		BIT(1)
+#define RNG_SR_SECS		BIT(2)
+#define RNG_SR_CEIS		BIT(5)
+#define RNG_SR_SEIS		BIT(6)
+
+#define RNG_TIMEOUT_US		1000
+
+struct stm32_rng_instance {
+	struct io_pa_va base;
+	unsigned long clock;
+};
+
+static struct stm32_rng_instance *stm32_rng;
+
+/*
+ * Extract from the STM32 RNG specification:
+ *
+ * When a noise source (or seed) error occurs, the RNG stops generating
+ * random numbers and sets to “1” both SEIS and SECS bits to indicate
+ * that a seed error occurred. (...)
+
+ * The following sequence shall be used to fully recover from a seed
+ * error after the RNG initialization:
+ * 1. Clear the SEIS bit by writing it to “0”.
+ * 2. Read out 12 words from the RNG_DR register, and discard each of
+ * them in order to clean the pipeline.
+ * 3. Confirm that SEIS is still cleared. Random number generation is
+ * back to normal.
+ */
+static bool conceal_seed_error(uintptr_t rng_base)
+{
+	if (read32(rng_base + RNG_SR) & (RNG_SR_SECS | RNG_SR_SEIS)) {
+		size_t i = 0;
+
+		io_mask32(rng_base + RNG_SR, 0, RNG_SR_SEIS);
+
+		for (i = 12; i != 0; i--)
+			(void)read32(rng_base + RNG_DR);
+
+		if (read32(rng_base + RNG_SR) & RNG_SR_SEIS)
+			panic("RNG noise");
+
+		return true;
+	}
+
+	return false;
+}
+
+TEE_Result stm32_rng_read_raw(uintptr_t rng_base, uint8_t *out, size_t size)
+{
+	uint8_t *buf = out;
+	size_t len = size;
+	TEE_Result rc = TEE_ERROR_SECURITY;
+	int count = 0;
+
+	/* Enable RNG if not, clock error is disabled */
+	if (!(read32(rng_base + RNG_CR) & RNG_CR_RNGEN))
+		write32(RNG_CR_RNGEN | RNG_CR_CED, rng_base + RNG_CR);
+
+	while (len) {
+		uint64_t timeout_ref = timeout_init_us(RNG_TIMEOUT_US);
+
+		/* Wait RNG has produced random well seeded samples */
+		while (!timeout_elapsed(timeout_ref)) {
+			if (!conceal_seed_error(rng_base) &&
+			    read32(rng_base + RNG_SR) & RNG_SR_DRDY)
+				break;
```

Comment:
This will only break the inner loop. How about adding an `out:` label below, and set `rc` here as needed and `goto out`?

Reply:
This is intended to break the inner loop and move to the sequence below:
- conceal seed error, if occurrence, wait again for samples.
- test if data are ready, if not, then it's an effective timeout.

The current sequence allows to conceal with seed errors without waiting for the timeout to do so.

Reply:
This is confusing. So if there's a conceal seed error we test again until it has vanished and data is ready. The first thing we do if data is ready is to test for conceal seed error and check if data is ready. If there's a timeout we check if the there's a conceal error, if so we restart the timeout and loop again.

The only way to get a timeout is if there isn't a conceal seed error but data isn't ready.

Is this a correct summary of intended behavior?

Reply:
hmmm... i'm confused myself :(
I start adding some mutex for RNG accesses, using spinlocks. This makes the loop uninterruptible, hence simplifies the timeout detection part. I will push an update once I have something.

I plan to make the `stm32_rng_read_raw()`  accessing RNG by 16 bytes data burst in atomic context, and `stm32_rng_read()` to loop,  locking/unlocking RNG for each 16byte max burst.